portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -35,22 +35,33 @@
   private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
   private static final String TIME = "time";
   private static final String NEWEST_MESSAGE = "newest-message";
+  private static final String COMMENT_LIMIT = "comment-limit";
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String limitString = request.getParameter("numComments");
+    int limit = convertToInt(limitString);
+
     Query query = new Query("Message").addSort("time", SortDirection.DESCENDING);
 
     PreparedQuery results = datastore.prepare(query);
 
+    List<String> limitedList = new ArrayList<>();
     List<String> messages = new ArrayList<>();
     for (Entity entity : results.asIterable()) {
-      String fullMessage = (String) entity.getProperty("newest-message");
-      long timestamp = (long) entity.getProperty("time");
+      String fullMessage = (String) entity.getProperty(NEWEST_MESSAGE);
+      long timestamp = (long) entity.getProperty(TIME);
 
       messages.add(fullMessage);
     }
 
-    String json = gson.toJson(messages);
+    limit = convertToInt(limitString);
+
+    for (int i = messages.size() - 1; i > messages.size() - 1 - limit; i--) {
```

Comment:
An elegant solution could be:

```java
List<String limitedList = 
    Lists.reverse(messages)
        .stream()
        .limit(limit)
        .collect(Collectors.toList();
```

1. `Lists.reverse` is creating a reversed list
2. We then use the Java stream API on the reversed list
3. We use the Stream API to keep the first `limit` items
4. We create a list from the result